api/wft-complete-metadata.md

```diff
@@ -0,0 +1,133 @@
+* Metrics work great for all sorts of latency measurements, but what if you wanted to know about
+  how long a _specific_ WFT took? You could emit that.
+* Interceptors could pipe whatever arbitrary data the user wants into there, for use cases where
+  memo / search attributes / side effects may not be appropriate.
+
+### Things we haven't thought of
+
+Having this metadata be suitably generic means we dont need to update the API repo for every
+possible usecase. Particularly ones where the SDK is the only one who needs to be able to read it.
+
+## Implementation Proposal
+
+We'd add a new metadata field to the
+[WFT response](https://github.com/temporalio/api/blob/master/temporal/api/workflowservice/v1/request_response.proto#L274).
+And the [WFT completed event](https://github.com/temporalio/api/blob/f0350f8032ad2f0c60c539b3b61ea37f412f1cf7/temporal/api/history/v1/message.proto#L186).
+Though the completion event need not necessarily contain all the info sent in the completion. If
+it's known and understood that a particular metadata field is simply telling server something
+ephemeral, that could be stripped before writing the event.
+
+With the following proto as its type:
+
+```protobuf
+message StructuredMetadata {
+  // For JSON-esque data representations that can be easily decoded by anyone. Could contain user
+  // data that they'd like to show up the UI, for example.
+  map<string, google.protobuf.Struct> struct_metadata = 1;
+  // For better typed / more efficient binary serialization of types that are expected to be known
+  // by their consumers.
+  map<string, google.protobuf.Any> proto_metadata = 2;
+}
+```
+
+Having both fields seems useful here, since `Any` is useful for the SDK making notes to itself which
+it knows the type of. `Struct` is essentially just a more efficient JSON representation, and can
+be used for arbitrary data that can easily be rendered.
+
+Since `Any` can also be `Struct`, we could have just one field, but this wastes some bytes on the
+`type_url` field if we are throwing a lot of structs in there. Keeping them separate also makes
+it clear what you can decode without any type knowledge, rather than walking the whole map looking
+for struct types in case you want to decode them - that way you only look in the `Any` map when you
+know you're looking for something specific.
+
+### Why not just add more fields to WFT completion response?
+
+* Things that the SDK only needs to write down for itself (ex: the versioning usecase) can now be
+  made without requiring changes to the API repo, updating the server, releasing it, and ensuring
+  that that version is in use. That's real nice.
+* Supports arbitrary user-originated data (TBD if we actually want to allow this, but with this
+  approach it's easy to add if we do).
+
+### Why not use markers?
+
+This is mentioned above, but I want to reiterate it: 
+
+I think it's very important that history events stay focused on recreating workflow state. The
```

Comment:
We're using this mechanism for versioning the SDK logic which effectively means it's for recreating workflow state.